torch/onnx/experimental/fonnx/v1/exporter.py

```diff
@@ -0,0 +1,116 @@
+    """Export a torch.nn.Module to a ONNX graph
+
+    Args:
+        model: Model to be exported
+        model_args: Positional inputs to `torch_model`
+        export_options: Export options
+        model_kwargs: Keyword inputs to `torch_model`
+
+    Returns:
+        An in-memory representation of the exported ONNX model.
+        See ``ExportOutput`` for reference.
+    """
+    ...
+
+
+class ExportLogLevel(Enum):
+    """Class for choosing log level.
+
+    Names and values follows https://docs.python.org/3/library/logging.html#levels"""
+
+    CRITICAL = 50
+    ERROR = 40
+    WARNING = 30
+    INFO = 20
+    DEBUG = 10
+    NOTSET = 0
+
+
+@dataclasses.dataclass
+class ExportOptions:
+    """Class for keeping track of ONNX exporter flags can perform."""
+
+    opset_version: Optional[int] = None
+    """opset_version: Target ONNX opset version
+
+    When not specified, the exporter selects the latest supported version."""
+
+    _log_level: Optional[Union["ExportLogLevel", str, int]] = None
+    """Log level as defined by https://docs.python.org/3/library/logging.html#levels"""
+
+    @property
+    def log_level(self) -> ExportLogLevel:
+        return self._log_level
+
+    @log_level.setter
+    def log_level(self, log_level: Optional[Union[ExportLogLevel, str, int]]):
+        # do the mapping to self._log_level, raise on invalid values
+        ...
+
+    dynamic_shape: Optional[bool] = None
+    """Hints shape information for input/output tensors.
+
+    When True, all inputs shapes are considered dynamic.
+    When False, all inputs shapes are considered static.
+    When not specified, the exporter sets to the most compatible setting.
+    """
+
+
+class ExportOutputSerializer(Protocol):
+    """Serializes ONNX graph into a specific format.
+
+    This is a base class that must be specialized with the desire format specification
+    """
+
+    def serialize(self, export_output: ExportOutput, destination: io.BytesIO) -> None:
+        pass
+
+
+class ProtobufExportOutputSerializer:
+    """Serializes ONNX graph as Protobuf."""
+
+    def serialize(self, export_output: ExportOutput, destination: io.BytesIO) -> None:
+        destination.write(export_output.onnx_model.SerializeToString())
+
+
+class ExportOutput:
+    """In-memory representation of a torch.nn.Module exported to ONNX."""
+
+    def save(
+        self,
+        destination: Union[str, io.BytesIO],
+        serializer: Optional[ExportOutputSerializer] = None,
+    ):
+        """Saves in-memory ONNX representation to `destination` using specified `serialer`"""
+
+        if serializer is None:
+            serializer = ProtobufExportOutputSerializer()
+        if isinstance(destination, str):
+            with open(destination, "wb") as f:
+                serializer.serialize(self, f)
+        else:
+            serializer.serialize(self, destination)
+
+    def to_bytes(
+        self, serializer: Optional[ExportOutputSerializer] = None
+    ) -> io.BytesIO:
```

Comment:
```suggestion
    ) -> bytes:
```

or is returning `io.BytesIO` intended? (the code below is returning `bytes`)

Reply:
probably `io.BytesIO` because that is what `onnx.load` would expect, right?

If the user wants to save the onnx, they would use the `ExportOutput.save` directly. If they are using `ExportOutput.to_bytes`, they are probably want to load the model as ONNX for unit testing?

In fact, if the *only* reason for `to_bytes` is unit testing, maybe it should be implemented as a private method or as a helper somewhere else. @BowenBao can you think of another use case for `to_bytes`?

Reply:
discussed offline. removed `to_bytes`